### PR TITLE
Adds macros for generating case class Parseables

### PR DESCRIPTION
**Pull request by gregghz**
Three macros are added, each has the same goal of automatically deriving
a `Parseable[A]` from a case class.

A common pattern is to define a case class and then define a
`Parseable[A]` for that case class. You end up writing code like this:

```
case class Foo(a: String, b: Int, c: Option[Date])
object Foo {
  implicit val parseable: Parseable[Foo] = new Parseable[Foo] {
    def parse(row: SqlResult): Foo = {
      Foo(row.string("a"), row.int("b"), row.dateOption("c"))
    }
  }
}
```

The amount of boilerplate increases proportionally with the number of
case class parameters.

This commit allows you to write the above as follows:

```
case class Foo(a: String, b: Int, c: Option[Date])
object Foo {
  implicit val parseable: Parseable[Foo] = generateParseable[Foo]
}
```

or:

```
@Record()
case class Foo(a: String, b: Int, c: Option[Date])
```

Both of these produce the same code as seen in the first example.

Sometimes your field names might not map one-to-one with your database
column names. A common pattern at Lucid is to use snake_case for column
names in the database while using camelCase for members of case classes.

In these cases you can still use a macro as follows:

```
case class User(firstName: String, lastName: String)
object Foo {
  implicit val parseable: Parseable[Foo] = generateSnakeParseable[Foo]
}
```

or

```
@Record("snakeCase" -> true)
case class User(firstName: String, lastName: String)
```

For more control over column names and to prevent changes in a case
class from breaking SQL queries, you also have the option to specify a
mapping of case class fields to column names:

```
case class User(firstName: String, lastName: String)
object Foo {
  implicit val parseable: Parseable[Foo] = generateParseable[Foo](
    "firstName" -> "fname",
    "lastName" -> "lname"
  )
}
```

or:

```
@Record("colMapping" -> Map(
  "firstName" -> "fname",
  "lastName" -> "lname"
))
case class User(firstName: String, lastName: String)
```

See the tests for complete examples (including examples of things that
should not compile).

### CODE REVIEW

**Comment by gregghz** [comment]
This adds a sub-project of relate. I want to publish this separately as well so that when you depend on "relate" you don't get the macros, but instead you have to also depend on "relate-macros". I don't know how to do this. @pauldraper might?

We also need macros/test running in travis. Again, I'm not sure how to do that.

**Comment by pauldraper** [comment]
> I want to publish this separately

Split out the common publishing settings, and make sure to apply them to both projects (e.g. `pgpPublicRing`).

Then either

1. Change the `after_success` section of .travis.yml to include `macro/publishSigned`.

1. Make an aggregate project. Actually, I'd recommend moving relate code to a `relate` subdirectory. Then the default project will automatically be an aggregate project that includes `relate` and `macro`.

> We also need macros/test running in travis.

Add `macros/test` to the sbt command in `script` section of .travis.yml.

File: build.sbt [hunk 1]
```diff
@@ -74,3 +78,19 @@ lazy val relate = Project(
...
 ).configs(Benchmark).settings(inConfig(Benchmark)(Defaults.testSettings): _*).
   configs(Regression).settings(inConfig(Regression)(Defaults.testSettings): _*)
+
+lazy val macroSettings = buildSettings ++ Seq(
+  moduleName := "relate-macros",
+  scalacOptions += "-language:experimental.macros",
+  addCompilerPlugin("org.scalamacros" % "paradise" % "2.1.0" cross CrossVersion.full),
+  libraryDependencies <+= (scalaVersion)("org.scala-lang" % "scala-reflect" % _),
+  libraryDependencies ++= Seq(
+    "com.chuusai" %% "shapeless" % "2.3.1" % "test",
```
Comment (pauldraper) [inline build.sbt]: He he.

Comment (msiebert) [inline build.sbt]: :fearful: shapeless... I'm gonna come back to this code base later, and I won't be able to understand it anymore...

Comment (gregghz) [inline build.sbt]: just in tests. `illTyped()` is a great function for testing for compilation failure :)

**Comment by gregghz** [comment]
@pauldraper I fixed the build to support scala 2.10.x. I had to include more code from the shapeless guy: https://github.com/milessabin/macro-compat.

**Pull request merged**